Writes up steps for running multiple pyrsia nodes on a macosx (#553)
pyrsia/pyrsia: docs/setup_on_macosx.md

# How to run 2 pyrsia nodes natively on 1 MacOS machine

​​Download a fresh copy of the codebase by cloning the repo or updating to the HEAD of `main`
Let's call this folder PYRSIA_HOME. We will refer to this name in the following steps.

Build binaries for pyrsia_node
`cargo build --workspace`

## Create 2 separate nodes 'installations'

> This will create two copies of the same binary so that you can configure them as independent nodes

Reviewer: ```suggestion
This will create two copies of the same binary so that you can configure them as independent nodes.

```

Create Node A
```
mkdir nodeA

cp target/debug/pyrsia_node nodeA
```
Create Node B
```
mkdir nodeB

cp target/debug/pyrsia_node nodeB
```
### Start Node A

> In a new terminal start node A, http listen on 7788 p2p listen on 44001

Reviewer: ```suggestion
In a new terminal start node A, http listen on 7888 and p2p listen on 44001
```

```
cd nodeA
DEV_MODE=on RUST_LOG="pyrsia=debug,info"  ./pyrsia_node -H 0.0.0.0 -p 7888 -L /ip4/0.0.0.0/tcp/44001
```

If everything goes well, you will see a line similar to the following in the logs on the terminal(The IP address could be different than in the sample below)
```
# INFO  pyrsia::network::p2p > Local node is listening on "/ip4/192.168.0.110/tcp/44001/p2p/12D3KooWLKMbBzp4k1mcM2rYXs8VQgoCSNLxGUwnB1itouxYcnx3"
```
If you do not find this line right away try running `grep 44001`

### Start Node B

> In a new terminal start node B, http listen on 7788 p2p listen on 44002 and connect to peer nodeA

Reviewer: ```suggestion
In a new terminal start node B, http listen on 7889, p2p listen on 44002 and connect to peer node A on port 44001
```

```
cd nodeB

> DEV_MODE=on RUST_LOG="pyrsia=debug,info"  ./pyrsia_node -H 0.0.0.0 -p 7889 -L /ip4/0.0.0.0/tcp/44002 --peer /ip4/127.0.0.1/tcp/44001

Reviewer 2: In the file [local_dev_setup.md](https://github.com/pyrsia/pyrsia/blob/main/docs/local_dev_setup.md), it mentions that Node 2 is listening on 8181, it is better to keep them consistent.

Author: Good point. We will leave this one here and I will update the `local_dev_setup.md` with the same port number `7889`

```

> If everything goes well, you will see a line similar to the following in the logs on the terminal(The IP address could be different than in the sample below)

Reviewer: ```suggestion
If everything goes well, you will see a line similar to the following in the logs on the terminal. (The IP address could be different than in the sample below)
```

### Good news
If you saw the above lines in your logs and did not see any failure/error messages your Pyrsia node network has now been setup.
Also that means we did not break anything ;P.

You should now be able to interact with the Pyrsia node.

## Using the Pyrsia CLI
On a fresh terminal, navigate to PYRSIA_HOME. We will now configure the pyrsia_cli to connect with Node B.
```
cd target/debug
./pyrsia config --add
Enter host:
localhost
Enter port:
7889
Enter disk space to be allocated to pyrsia(Please enter with units ex: 10 GB):
```
If everything worked well, you will see the following success message.
```
Node configuration Saved !!
```

Now you are ready to use the Pyrsia CLI.

Let us run through a few examples of how you can use the Pyrsia CLI
### Get Node status
```
./pyrsia node -s
Connected Peers Count:       1
Artifacts Count:             0
Total Disk Space Allocated:  10.84 GB
Disk Space Used:             0.0000%
```

### List all known peers
```
./pyrsia node -l
Connected Peers:
["12D3KooWH1tJB9NMuzHcEd6TU9yG4mv2Lo4J2gaXaBLpyNCrqRR9"]
```

Now that you have setup both the Pyrsia Node and Pyrsia CLI you are ready to start using Pyrsia.

# Using Pyrsia with Docker

> Once you have setup the pyrsia nodes and the cli you are ready to start using Pyrsia with docker.

Reviewer: ```suggestion
Once you have setup the pyrsia nodes and the CLI you are ready to start using Pyrsia with docker.
```

> ## Configure docker desktop to use nodeA as registry mirror

Reviewer: ```suggestion
## Configure docker desktop to use node A as registry mirror
```

In your Docker desktop installation -> Settings -> Docker Engine where docker allows you to set registry-mirrors.

> Setup nodeA as a registry mirror by adding/editing the following in the configuration

Reviewer: ```suggestion
Setup node A as a registry mirror by adding/editing the following in the configuration.
```

```
 "registry-mirrors": [
   "http://192.168.0.110:7888" (IP address of host machine: port number for node A)
 ]
```
On MacOSX using localhost does not work(because the request is made from the Docker Desktop VM), so you will need to specify the IP address of host machine

You will need to restart Docker Desktop.
Once restarted you should be able to pull docker images through Pyrsia

> # pull alpine and make sure it's not in local docker cache:

Reviewer: ```suggestion
## pull alpine

first make sure alpine is not in local docker cache, then pull alpine:
```

```
docker rmi alpine # remove alpine from local docker cache
docker pull alpine
```
When you pull an image of alpine from dockerhub Pyrsia node A should act as a pull-through cache and show a line similar to the following in its log
```
# DEBUG pyrsia::docker::v2::handlers::blobs> Step 3: "sha256:3d243047344378e9b7136d552d48feb7ea8b6fe14ce0990e0cc011d5e369626a" successfully stored locally from docker.io
```

> You can try the same with node B acting as the registry mirros

Reviewer: ```suggestion
You can try the same with node B acting as the registry mirror.

```

Change the docker registry mirror to node B
```
 "registry-mirrors": [
   "http://192.168.0.110:7889" (IP address of host machine: port number of node B)
 ]
```
Remove alpine and perform a docker pull again
```
docker rmi alpine # remove alpine from local docker cache
docker pull alpine
```

> Now node B is acting as the pull-through cache and should show a line similar to the following,  in its log

Reviewer: ```suggestion
Now node B is acting as the pull-through cache and should show a line similar to the following in its log, indicating `alpine` was retrieved from the Pyrsia network (in this case node A).
```
